File: crates/aip-core/src/keys.rs
//! Key management for AIP identities.
//!
//! # Security
//!
//! This module handles secret key material. Key security properties:
//!
//! - **Zeroization on drop**: Secret key bytes are automatically zeroed when
//!   dropped. This is handled by `ed25519-dalek`'s `SigningKey` via its
//!   `zeroize` feature, preventing leakage via memory dumps, swap files,
//!   or cold boot attacks.
//!
//! - **No Debug leakage**: Keys implement Debug safely, showing only the
//!   DID (public info), not secret material.

use crate::{Did, Error, Result};
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use rand::rngs::OsRng;
use std::fmt;

/// A root identity key.
///
/// This is the primary key that defines an agent's identity.
/// It should be stored securely and used sparingly.
///
/// # Security
///
/// - The inner `SigningKey` automatically zeroizes secret key bytes on drop
///   (via ed25519-dalek's `zeroize` feature).
/// - The `Debug` implementation only shows the DID (public info) to prevent
///   accidental exposure of key material in logs.
pub struct RootKey {
    signing_key: SigningKey,
}

impl fmt::Debug for RootKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("RootKey")
            .field("did", &self.did().to_string())
            .finish_non_exhaustive()
    }
}

impl RootKey {
    /// Generate a new random root key.
    pub fn generate() -> Self {
        Self {
            signing_key: SigningKey::generate(&mut OsRng),
        }
    }

    /// Create from existing bytes.
    ///
    /// # Security
    ///
    /// The caller should zeroize the source bytes after this call
    /// if they are no longer needed.
    pub fn from_bytes(bytes: &[u8; 32]) -> Result<Self> {
        Ok(Self {
            signing_key: SigningKey::from_bytes(bytes),
        })
    }

    /// Get the DID for this root key.
    pub fn did(&self) -> Did {
        Did::new(self.signing_key.verifying_key())
    }

    /// Get the public verifying key.
    pub fn verifying_key(&self) -> VerifyingKey {
        self.signing_key.verifying_key()
    }

    /// Sign a message.
    pub fn sign(&self, message: &[u8]) -> Signature {
        self.signing_key.sign(message)
    }

    /// Get the secret key bytes.
    ///
    /// # Security Warning
    ///
    /// This returns raw secret key material. The caller is responsible for:
    /// - Storing the bytes securely
    /// - Zeroizing the bytes when no longer needed
    /// - Not logging or printing these bytes
    pub fn to_bytes(&self) -> [u8; 32] {
        self.signing_key.to_bytes()
    }
}

/// A session key delegated from a root key.
///
/// Used for day-to-day operations without exposing the root key.
///
/// # Security
///
/// - The inner `SigningKey` automatically zeroizes secret key bytes on drop.
/// - Session keys should be short-lived and rotated frequently.
/// - The `Debug` implementation only shows public info (root DID and pubkey
///   fingerprint).
pub struct SessionKey {
    signing_key: SigningKey,
    root_did: Did,
}

impl fmt::Debug for SessionKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Show a short fingerprint of the public key (first 8 chars of base58)
        let pubkey_fingerprint = {
            let full = self.public_key_base58();
            if full.len() > 8 {
                format!("{}...", &full[..8])
            } else {
                full
            }
        };

        f.debug_struct("SessionKey")
            .field("root_did", &self.root_did.to_string())
            .field("pubkey", &pubkey_fingerprint)
            .finish_non_exhaustive()
    }
}

impl SessionKey {
    /// Generate a new session key for a root identity.
    pub fn generate(root_did: Did) -> Self {
        Self {
            signing_key: SigningKey::generate(&mut OsRng),
            root_did,
        }
    }

    /// Get the root DID this session key belongs to.
    pub fn root_did(&self) -> &Did {
        &self.root_did
    }

    /// Get the public verifying key.
    pub fn verifying_key(&self) -> VerifyingKey {
        self.signing_key.verifying_key()
    }

    /// Sign a message.
    pub fn sign(&self, message: &[u8]) -> Signature {
        self.signing_key.sign(message)
    }

    /// Get the public key as base58.
    pub fn public_key_base58(&self) -> String {
        bs58::encode(self.signing_key.verifying_key().as_bytes()).into_string()
    }
}

/// Verify a signature against a public key.
pub fn verify(public_key: &VerifyingKey, message: &[u8], signature: &Signature) -> Result<()> {
    public_key
        .verify(message, signature)
        .map_err(|_| Error::InvalidSignature)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_root_key_generation() {
        let root = RootKey::generate();
        let did = root.did();
        assert!(did.to_string().starts_with("did:key:"));
    }

    #[test]
    fn test_sign_verify() {
        let root = RootKey::generate();
        let message = b"hello world";
        let signature = root.sign(message);

        verify(&root.verifying_key(), message, &signature).unwrap();
    }

    #[test]
    fn test_session_key() {
        let root = RootKey::generate();
        let session = SessionKey::generate(root.did());

        assert_eq!(session.root_did(), &root.did());

        let message = b"session message";
        let sig = session.sign(message);
        verify(&session.verifying_key(), message, &sig).unwrap();
    }

    #[test]
    fn test_root_key_roundtrip() {
        let root = RootKey::generate();
        let bytes = root.to_bytes();
        let restored = RootKey::from_bytes(&bytes).unwrap();

        assert_eq!(root.did(), restored.did());
    }

    #[test]
    fn test_root_key_debug_does_not_leak_secrets() {
        let root = RootKey::generate();
        let debug_output = format!("{:?}", root);

        // Should contain the DID (public info)
        assert!(debug_output.contains("did:key:"));

        // Should NOT contain any of these patterns that would indicate leaked key material
        assert!(
            !debug_output.contains("FieldElement"),
            "Debug output should not contain internal crypto field elements"
        );
        assert!(
            !debug_output.contains("EdwardsPoint"),
            "Debug output should not contain internal crypto types"
        );
        assert!(
            !debug_output.to_lowercase().contains("secret"),
            "Debug output should not reference 'secret'"
        );

        // Should use finish_non_exhaustive (indicated by "..")
        assert!(
            debug_output.contains(".."),
            "Debug should indicate hidden fields with .."
        );
    }

    #[test]
    fn test_session_key_debug_does_not_leak_secrets() {
        let root = RootKey::generate();
        let session = SessionKey::generate(root.did());
        let debug_output = format!("{:?}", session);

        // Should contain root_did and truncated pubkey
        assert!(debug_output.contains("root_did"));
        assert!(debug_output.contains("pubkey"));

        // Pubkey should be truncated (ends with ...)
        assert!(
            debug_output.contains("...\""),
            "Public key should be truncated in debug output"
        );

        // Should NOT contain leaked key material
        assert!(
            !debug_output.contains("FieldElement"),
            "Debug output should not contain internal crypto field elements"
        );
        assert!(
            !debug_output.contains("EdwardsPoint"),
            "Debug output should not contain internal crypto types"
        );
        assert!(
            !debug_output.to_lowercase().contains("secret"),
            "Debug output should not reference 'secret'"
        );
    }
}

File: sdk/python/agent_id/keys.py
"""
Key management for Agent Identity Protocol.

Provides RootKey (long-lived identity key) and SessionKey (short-lived delegated key).
Uses Ed25519 for all cryptographic operations.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from nacl.signing import SigningKey, VerifyKey
from nacl.encoding import RawEncoder

from agent_id.did import Did


@dataclass
class RootKey:
    """
    The root identity key for an agent.
    
    This is the long-lived key that defines the agent's identity (DID).
    Should be stored securely and used sparingly - prefer session keys
    for routine operations.
    
    Attributes:
        signing_key: The Ed25519 signing key (private)
        verify_key: The Ed25519 verify key (public)
        did: The agent's DID derived from the public key
    """
    
    signing_key: SigningKey
    verify_key: VerifyKey
    did: Did
    
    @classmethod
    def generate(cls) -> RootKey:
        """
        Generate a new random root key.
        
        Returns:
            A new RootKey with a fresh Ed25519 keypair
        """
        signing_key = SigningKey.generate()
        verify_key = signing_key.verify_key
        public_key_bytes = bytes(verify_key)
        did = Did.from_public_key(public_key_bytes)
        
        return cls(
            signing_key=signing_key,
            verify_key=verify_key,
            did=did,
        )
    
    @classmethod
    def from_seed(cls, seed: bytes) -> RootKey:
        """
        Create a root key from a 32-byte seed.
        
        Args:
            seed: 32-byte seed for deterministic key generation
            
        Returns:
            A RootKey derived from the seed
            
        Raises:
            ValueError: If seed is not 32 bytes
        """
        if len(seed) != 32:
            raise ValueError(f"Seed must be 32 bytes, got {len(seed)}")
        
        signing_key = SigningKey(seed)
        verify_key = signing_key.verify_key
        public_key_bytes = bytes(verify_key)
        did = Did.from_public_key(public_key_bytes)
        
        return cls(
            signing_key=signing_key,
            verify_key=verify_key,
            did=did,
        )
    
    @classmethod
    def from_bytes(cls, private_key_bytes: bytes) -> RootKey:
        """
        Create a root key from raw private key bytes.
        
        Args:
            private_key_bytes: 32-byte Ed25519 private key
            
        Returns:
            A RootKey from the provided private key
        """
        return cls.from_seed(private_key_bytes)
    
    def sign(self, message: bytes) -> bytes:
        """
        Sign a message with this root key.
        
        Args:
            message: The message bytes to sign
            
        Returns:
            64-byte Ed25519 signature
        """
        signed = self.signing_key.sign(message, encoder=RawEncoder)
        return signed.signature
    
    def to_bytes(self) -> bytes:
        """
        Export the private key as bytes.
        
        Warning: Handle with care - this is the root identity!
        
        Returns:
            32-byte private key
        """
        return bytes(self.signing_key)
    
    def __repr__(self) -> str:
        return f"RootKey(did={self.did})"


@dataclass
class SessionKey:
    """
    A short-lived session key delegated from a root key.
    
    Session keys are used for routine operations to limit exposure
    of the root key. They should be rotated frequently.
    
    Attributes:
        signing_key: The Ed25519 signing key (private)
        verify_key: The Ed25519 verify key (public)
        root_did: The DID of the root key that delegated this session
        key_id: Identifier for this session key (e.g., "session-1")
    """
    
    signing_key: SigningKey
    verify_key: VerifyKey
    root_did: Did
    key_id: str
    
    @classmethod
    def generate(cls, root_key: RootKey, key_id: Optional[str] = None) -> SessionKey:
        """
        Generate a new session key delegated from a root key.
        
        Args:
            root_key: The root key to delegate from
            key_id: Optional identifier for this session key
            
        Returns:
            A new SessionKey
        """
        import secrets
        
        signing_key = SigningKey.generate()
        verify_key = signing_key.verify_key
        
        if key_id is None:
            key_id = f"session-{secrets.token_hex(4)}"
        
        return cls(
            signing_key=signing_key,
            verify_key=verify_key,
            root_did=root_key.did,
            key_id=key_id,
        )
    
    def sign(self, message: bytes) -> bytes:
        """
        Sign a message with this session key.
        
        Args:
            message: The message bytes to sign
            
        Returns:
            64-byte Ed25519 signature
        """
        signed = self.signing_key.sign(message, encoder=RawEncoder)
        return signed.signature
    
    @property
    def full_key_id(self) -> str:
        """The full key ID including the root DID."""
        return f"{self.root_did}#{self.key_id}"
    
    def __repr__(self) -> str:
        return f"SessionKey(root={self.root_did}, id={self.key_id})"


def verify_signature(
    message: bytes,
    signature: bytes,
    public_key: bytes,
) -> bool:
    """
    Verify an Ed25519 signature.
    
    Args:
        message: The original message bytes
        signature: The 64-byte signature
        public_key: The 32-byte public key
        
    Returns:
        True if signature is valid, False otherwise
    """
    try:
        verify_key = VerifyKey(public_key)
        verify_key.verify(message, signature)
        return True
    except Exception:
        return False
